wlp-managed-8.5/src/main/java/org/jboss/arquillian/container/was/wlp_managed_8_5/WLPManagedContainer.java

```diff
@@ -340,23 +391,81 @@ public ProtocolMetaData deploy(final Archive<?> archive) throws DeploymentExcept
          throw new DeploymentException("Exception while deploying application.", e);
       }
    }
-   
+
+   private List<String> findArquillianContextRoots(final EnterpriseArchive ear, String deployName) throws DeploymentException {
+	   List<String> contextRoots = new ArrayList<String>();
+	   int testableWarCounter = 0;
+	   int totalWarCounter = 0;
+	   WebArchive latestWar = null;
+	   for (ArchivePath path : ear.getContent().keySet()) {
+		   if (path.get().endsWith("war")) {
+			   WebArchive war = ear.getAsType(WebArchive.class, path);
+			   totalWarCounter++;
+			   if (Testable.isArchiveToTest(war)) {
+				   contextRoots.add(getContextRoot(ear, war));
+				   testableWarCounter++;
+			   }
+			   latestWar = war;
+		   }
+	   }
+	   if(testableWarCounter == 0) {
+		   if(totalWarCounter == 1) { // fallback only one war
+			   contextRoots.add(getContextRoot(ear, latestWar));
+		   } else { // default fallback
+			   contextRoots.add(deployName);
+		   }
+	   }
+	   return contextRoots;
+	}
+
+   private String getContextRoot(EnterpriseArchive ear, WebArchive war) throws DeploymentException {
+	   org.jboss.shrinkwrap.api.Node applicationXmlNode = ear.get("META-INF/application.xml");
+	   if(applicationXmlNode != null && applicationXmlNode.getAsset() != null) {
+		   InputStream input = null;
+		   try {
+			   input = ear.get("META-INF/application.xml").getAsset().openStream();
+			   Document applicationXml = readXML(input);
+			   XPath xPath = XPathFactory.newInstance().newXPath();
+			   XPathExpression ctxRootSelector = xPath.compile("//module/web[web-uri/text()='"+ war.getName() +"']/context-root");
+			   String ctxRoot = ctxRootSelector.evaluate(applicationXml);
+			   if(ctxRoot != null && ctxRoot.trim().length() > 0) {
+				   return ctxRoot;
+			   }
+		   } catch (Exception e) {
+			   throw new DeploymentException("Unable to delete archive from dropIn directory");
```

Comment:
That exception message needs to be adjusted to indicate what we're trying to do here.